### PR TITLE
REMARK version.2™

### LINKED ISSUE #183: missing royalty
```json
{
  "collection": "b8fdf4d6640c2aa356-FARM_WAIFU",
  "symbol": "CORN_WAIFU_2",
  "transferable": 1,
  "sn": "00000002",
  "metadata": "ipfs://ipfs/QmXHAwFWZnfhC4eDa6vwsscBBRrF19MVdrM9PrCzVdRVVe",
  "properties": {
    "royaltyInfo": {
      "type": "royalty",
      "value": {
        "receiver": "GksmaSfxzmD2Bxh7MZBTyGXTnGrQwoKnvukekh7JVzD8RVC",
        "royaltyPercentFloat": 20
      }
    }
  }
}
```

### PR DESCRIPTION
**Pull request by vikiival**
- :tada: v2 initial
- :package: META:squid:
- :zap: use MetaSquid
- :truck: updated imports from metasquid
- :zap: versioned router
- :top: :package: metasquid@0.1.0
- :heavy_plus_sign: @kodadot1/minipfs
- ga
- :label: just types :)))
- :truck: handle metadata moved to shared package
- :truck: common handler for create collection
- :zap: add version to the collection
- :zap: createCollection is a common friend
- :top: :package: metasquid@0.1.0
- :truck: moved createEvent
- :truck: moved burn into shared
- :truck: moved buy into shared
- :truck: moved change into shared
- :truck: moved emote into shared
- :truck: moved index into shared
- :truck: moved list into shared
- :truck: moved send into shared
- :truck: take all imports from shared
- :truck: moved createEvent to event
- :truck: moved mintNFT to separated file
- :alien: :soap: sort entity fields by name
- :alien: extended interactions with RMRK v2
- :construction: lock depends on #153
- :construction: adding remarkable entities
- :bug: emote should update nft updatedAt state
- @ma-shulgin dropped schema
- :construction: generated files
- :technologist: getRemark
- :tada: RMRK::BASE
- :tada: getters for V2
- :zap: ofloading more into future kodadot/metasquid
- :zap: createBaseId
- :bug: baseId will throw error on empty
- :zap: create base id
- :zap: max has value
- :alien: some more fields to base
- :alien: every entity now stores version
- :arrow_up: @subsquid deps
- :squid: typegen
- :see_no_evil: SQL Dumps
- :arrow_up: Other deps
- :arrow_down: nano-id is rekt at v4
- :squid: new formatting
- :necktie: BaseType
- :necktie: We only interact in V2
- :tada: Base is here
- :soap: order entities more correcly
- :truck: renaming like it is 90s
- :recycle: bit nicer v1 mint
- :truck: RmrkEvent is now Action
- :label: conditional types
- :label: proper types for V1
- :heavy_plus_sign: @vikiival/minimark
- :robot: quick
- :zap: add version to the mint
- :art: slowly to V2
- :test_tube: Some_tests and renames
- :bricks: using new minimark correctly
- :beers: Testing smoll RMRKv2 subset
- :bug: Test collection was created in block 16259864
- :wastebasket: .setBatchSize is obsolete
- :heavy_plus_sign: @vikiival/minimark
- :art: use interaction from v2 instead of action
- :loud_sound: more logs
- :arrow_up: @vikiival/minimark
- :label: put remark version
- :zap: event now remebers version
- :zap: LIST has version
- 🙄 reverting couple of commits ago
- :label: BaseCall has now RMRK version
- :zap: geeters with new unrwap
- :truck: renaming up and down
- :construction: minimal impl for acceptResource
- :bookmark: @vikiival/minimark@0.1.0-rc.7
- :broom: less own code, more minimark
- :zap: Base Should work
- 😈 co-pilot
- :zap: getAs is super smart
- :bookmark: @vikiival/minimark@0.1.0-rc.9
- :construction: miminimum of resource add
- :construction: Remarkable  Boilerplate
- :arrow_up: @vikiival/minimark
- :construction: set Priority and set property
- :truck: moved mint into shared
- :construction: Eqquipable
- 🙏 Zeit.
- :zap: send should work
- :card_file_box: version for Emote
- :tada: Emote has version
- :truck: rename of mintNFT to mintItem
- :zap: enable list
- :tada: burn is here
- :arrow_up: @subsquid deps
- :arrow_up: @vikiival/minimark
- :zap: mint should work
- :arrow_up: @polkadot/types
- :bug: faulty imports
- :card_file_box: nichts is fur dich
- :bug: V1 is not V1
- :construction: magical bug like wtf
- :bug:  Query runner already released. Cannot run queries anymore
- :robot: up in detached mode
- :bug: forgot collection hash
- :robot: bug always in detached mode
- :bug: RMRKv2 has symbol
- :truck: rename to addResource
- :bug: resource is pending when current owner is not issuer
- :alien: almost done schema for resource
- :zap: isIssuerOrError
- :zap: utils for #180
- :zap: dummyAddress
- :construction: Send V2
- :zap: assert not null on is Address
- :arrow_up: MetaSquid
- 🧠 RMRK team on the peak
- :alien: Resource <3
- :squid: Swim with squids
- :card_file_box: Resource and parent
- :arrow_up: MetaSquid
- :zap: recursive query to find root owner
- :arrow_up: MetaSquid
- :arrow_up: MetaSquid
- :construction: Send V2
- :mute: Deleted not fancy log
- :mute: Deleted not fancy log
- :bug: current_owner 101
- :arrow_up: MetaSquid
- :zap: hunger saved the rave
- :bug: SEND is using v2 fn
- :arrow_up: MetaSquid
- :card_file_box: pending for nestable nfts
- :robot: db
- :zap: nicely find root
- :zap: Wish that it works
- :zap: accept
- :broom: getOrFail
- :technologist: use get or fail to avoid ugly code
- :broom: remove unused imports
- :zap: addResource??
- :zap: base??
- :bug: not setting priority to resource
- :zap: setPriority??
- 🌚 something
- :zap: lock??
- 🎲 LET IT ROLL?!
- 🌈 something
- :wrench: Squid.yml
- :arrow_up: new squids
- :arrow_up: deps
- :arrow_up: @vikiival/minimark
- :tada: Ready for prod
- :zap: enable cachce
- :tada: start with Beta birds
- :arrow_up: @vikiival/minimark
- :bug: properly match remark
- :arrow_up: @kodadot1
- :card_file_box: royalty
- :arrow_up: @subsquid deps
- :broom: resorted fields for collection and nft
- :arrow_up: @kodadot1/metasquid@0.1.5-rc.0
- :bug: non-existing joins :)
- :zap: getRemarkString
- :broom: cleanup
- :construction: use metaquid logger instead of signale
- :wrench: smaller testable dataset
- :zap: new handy logging fns
- :loud_sound: in src/mappings/shared/burn.ts
- :loud_sound: in src/mappings/shared/buy.ts
- :loud_sound: in src/mappings/shared/change.ts
- :loud_sound: in src/mappings/shared/create.ts
- :loud_sound: in src/mappings/shared/emote.ts
- :loud_sound: in src/mappings/shared/list.ts
- :loud_sound: in src/mappings/shared/mint.ts
- :loud_sound: in src/mappings/shared/send.ts
- :wastebasked: removed star
- :zap: use pending and err
- :loud_sound: in src/mappings/v2/accept.ts
- :loud_sound: in src/mappings/v2/addResource.ts
- :loud_sound: in src/mappings/v2/base.ts
- :loud_sound: in src/mappings/v2/create.ts
- :loud_sound: in src/mappings/v2/equip.ts
- :loud_sound: in src/mappings/v2/equippable.ts
- :loud_sound: in src/mappings/v2/index.ts
- :loud_sound: in src/mappings/v2/lock.ts
- :loud_sound: in src/mappings/v2/mint.ts
- :loud_sound: in src/mappings/v2/send.ts
- :loud_sound: in src/mappings/v2/setpriority.ts
- :loud_sound: in src/mappings/v2/setproperty.ts
- :bug: logger residuals
- :arrow_up: MINIMARK :D
- :bug: no image for u
- :zap: tracking of elapsed time for metadata fetching
- 🌀 back to original dataset
- :zap: ability to fetch all metadata
- :zap: updated metadata every 10 minutes
- :tada: enable async cache updater
- :loud_sound: logs trace mint error
- 💀 disable metadata updater
- :robot: kek
- :mute: Cache logs are annoying
- :zap: logid
- :tada: lets cache
- :zap: one more time
- :card_file_box: symbol can be not unique
- :zap: enable series insight
- :bug: series insight to 100
- :bug: rollback cache
- :wrench: add royalty
- :card_file_box: royalty
- :arrow_up: MINIMARK
- :arrow_up: MINIMARK
- :zap: royalty closes #183
- :arrow_up: @kodadot1/minipfs@0.2.1
- :zap: get rmrk gateway
- ♻ replace @vikiival/minimark to @kodadot1/minimark
- ♻ replace in Files @vikiival/minimark to @kodadot1/minimark
- :arrow_up: @subsquid cli
- :wrench: :squid: yaml
- :wrench: set it to marck
- :zap: LIMIT cache to 20 per roundtrip
- :bug: Bad union
- :wrench: separated deployments for squids
- :robot: new squid setup
- :see_no_evil: build
- :robot: squid config justfile
- :arrow_up: @subsquid cli
- :wrench: dedicated true
- :robot: hard reset
- :bug: that is hanging my impl
- “Redeploy” is command to just restart containers
- :zap: validate transferability
- :bug: skip validating transferability on Unlist
- :ambulance: aaaaaaa
- :alien: ROYALTY EVENT
- :zap: royalty emits event
- :arrow_up: @kodadot1/minipfs
- :rocket: hyperspeed metadata
- :wrench: enable subscriptions
- :adhesive_bandage: Equip
- :memo: enable if you need V1
- :mute: Non-RMRK value is useless to log

### PR type

- [x] Bugfix
- [x] Feature
- [x] Refactoring

### What's new?

RMRK V2

### Optional

- [x] I've run the indexer and it hasn't failed
- [x] I've changed schema and performed a migration
- [x] I found edge cases - all defined in 
- #188 

### Screenshot

i

### CODE REVIEW

File: schema.graphql [hunk 1]
```diff
@@ -52,19 +38,25 @@ type NFTEntity @entity {
   meta: MetadataEntity
   metadata: String
   name: String @index
+  parent: NFTEntity # = nft_id
+  pending: Boolean!
```
Comment (vikiival) [inline schema.graphql:42]: - https://github.com/kodadot/rubick/issues/182 ?

File: schema.graphql [hunk 2]
```diff
@@ -108,21 +102,80 @@ type Emote @entity {
...
+    type: BaseType!
+    symbol: String!
+    issuer: String!
+    currentOwner: String!  
+    meta: MetadataEntity
+    metadata: String
+    # parts: [BasePart!] @derivedFrom(field: "base")
+    # themes: [BaseTheme!] @derivedFrom(field: "base")
+    # attachedToResources: [Resource!] @derivedFrom (field: "base")
+    # logs: [Call!] @derivedFrom (field: "base") // Rename to event
```
Comment (vikiival) [inline schema.graphql:146]: - #211

File: schema.graphql [hunk 3]
```diff
@@ -108,21 +102,80 @@ type Emote @entity {
...
 }
 
 
+# https://github.com/rmrk-team/rmrk-spec/blob/master/standards/rmrk2.0.0/entities/nft.md#resources-and-resource
+type Resource @entity {
+    id: ID!
+    # base: Base
```
Comment (vikiival) [inline schema.graphql:153]: - #210

File: src/mappings/shared/buy.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    const nft = await getWith<NFTEntity>(context.store, NFTEntity, interaction.id, { collection: true })
+    isInteractive(nft)
+    isPositiveOrElseError(nft.price, true)
+    if (version === '2.0.0') {
+      isMoreTransferable(nft, blockNumber)
+    }
+    isBuyLegalOrElseError(nft, extra || [])
+    const originalPrice = nft.price
+    const originalOwner = nft.currentOwner ?? undefined
+    nft.currentOwner = caller
```
Comment (vikiival) [inline src/mappings/shared/buy.ts:29]:
```
0x{bytes(RMRK::BUY::{version}::{id}::{recipient?})}
```

File: src/mappings/shared/change.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+import { CollectionEntity } from '../../model'
+import { unwrap } from '../utils'
+import { isOwnerOrElseError, withMeta } from '../utils/consolidator'
+import { getInteraction } from '../utils/getters'
+import { error, success } from '../utils/logger'
+import { Action, Context, RmrkInteraction } from '../utils/types'
+
+const OPERATION = Action.CHANGEISSUER
+
+// TODO: can also change BASE in V2 (not implemented yet)
```
Comment (vikiival) [inline src/mappings/shared/change.ts:14]: Todo still missing

File: src/mappings/shared/emote.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+import { unwrap } from '../utils'
+import { withMeta } from '../utils/consolidator'
+import { getInteraction } from '../utils/getters'
+import { emoteId } from '../utils/helper'
+import { error, success } from '../utils/logger'
+import { Action, Context, RmrkInteraction } from '../utils/types'
+
+const OPERATION = Action.EMOTE
+
+export async function emote(context: Context) {
```
Comment (vikiival) [inline src/mappings/shared/emote.ts:16]: missing namespace

```
RMRK::EMOTE::2.0.0::RMRK1::5105000-0aff6865bed3a66b-DLEP-DL15-00000001::1F389
```

```
0x{bytes(RMRK::EMOTE::{version}::{namespace}::{id}::{emotion})}
```

Screenshot 2023-04-04 at 15 17 21

Comment (vikiival) [reply src/mappings/shared/emote.ts:16]: https://kusama.subscan.io/extrinsic/0x4d0adddcfe4978560f5c6653d68d4ddf17703af8bcf750a2b5dd0189f502c0be

File: src/mappings/shared/mint.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    const collection = await get<CollectionEntity>(context.store, CollectionEntity, nft.collection)
+    isOwnerOrElseError(collection, caller)
+    const id = getNftId(nft, blockNumber)
+    // const entity = await get<NFTEntity>(context.store, NFTEntity, id) // TODO: check if exists
+    // plsNotBe<NFTEntity>(real, entity as NFTEntity)
+    const final = create<NFTEntity>(NFTEntity, id, {})
+    final.id = id
+    final.hash = md5(id)
+    final.issuer = caller
+    final.currentOwner = caller
```
Comment (vikiival) [inline src/mappings/shared/mint.ts:40]:
```
0x{bytes(rmrk::MINT::{version}::{html_encoded_json}::{recipient?})}
```

Missing recipient

File: src/mappings/utils/types.ts [hunk 8]
```diff
@@ -65,11 +67,27 @@ export interface IEvent<T = RmrkEvent> {
...
+type Theme = string | Record<string, string>
+type Themes = Record<string, Theme>
+
+export type Base = {
+  symbol: string
+  type?: BaseType
+  themes?: Themes
+  parts: any[]
+  metadata?: string
+}
```
Comment (vikiival) [inline src/mappings/utils/types.ts:90]: Is it in minimark?

File: src/mappings/v1/index.ts [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+import { updateCache } from '../utils/cache'
+
+import { burn as consume, buy, changeIssuer, createCollection, emote, list, send } from '../shared'
+import logger, { pending } from '../utils/logger'
+import {
+  Context,
+  RmrkInteraction
+} from '../utils/types'
+import { mintItem } from '../shared/mint'
+import { Interaction } from '@kodadot1/minimark/v1'
```
Comment (vikiival) [inline src/mappings/v1/index.ts:13]: Sort imports

File: src/mappings/v2/equip.ts [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+  try {
+    const { value: equip, caller, timestamp, blockNumber, version } = unwrap(context, getEquip);
+    interaction = equip
+    const nft = await get<NFTEntity>(context.store, NFTEntity, interaction.id)
+    plsNotBe(burned, nft)
+    isOwnerOrElseError(nft, caller)
+    nft.updatedAt = timestamp
+
+    // TODO: add logic for EQUIPing resource
```
Comment (Matehoo) [inline src/mappings/v2/equip.ts:27]: Will this be done in this PR?

File: src/mappings/v2/equippable.ts [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    const getE = getAs<Interaction.EQUIPPABLE>()
+    const { value: equip, caller, timestamp, blockNumber, version } = unwrap(context, getE);
+    interaction = equip
+    const nft = await get<NFTEntity>(context.store, NFTEntity, interaction.id)
+    plsBe(real, nft)
+    plsNotBe(burned, nft)
+    isOwnerOrElseError(nft, caller)
+    nft.updatedAt = timestamp
+
+    // TODO: add logic for EQUIPing resource
```
Comment (Matehoo) [inline src/mappings/v2/equippable.ts:29]: Will this be done in this PR?

File: src/mappings/v2/setproperty.ts [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+export async function setProperty(context: Context) {
+  let interaction: Optional<SetProperty> = null
+
+  try {
+    const getter = getSetProperty
+    const { value: interaction, caller, timestamp, blockNumber, version } = unwrap(context, getter);    
+
+    // logger.success(`[${OPERATION}] NEW PRIORITY ${interaction.value} for ${nft.id} from ${caller}`)
+
+    // TODO: add logic for accepting resource
```
Comment (Matehoo) [inline src/mappings/v2/setproperty.ts:20]: Will this be done in this PR?

**Review by Matehoo — COMMENTED** [review]
Some questions and nitpicks, LGTM 👍

File: src/mappings/v2/create.ts [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+import { handleMetadata } from '../shared'
+import { unwrap } from '../utils/extract'
+import { error, success } from '../utils/logger'
+import { Action, Collection, Context } from '../utils/types'
+import { getCreateCollection } from './getters'
+
+const OPERATION = Action.CREATE
+
+
+export async function createCollection(context: Context): Promise<void> {
```
Comment (vikiival) [inline src/mappings/v2/create.ts:16]: What is the diff between shared and this one?

File: src/mappings/v2/base.ts [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+    const { value: interaction, caller, timestamp, blockNumber, version } = unwrap(context, getCreateBase);
+    const base = interaction.value as CreatedBase
+    const id = baseId(blockNumber, base.symbol)
+    const final = await createUnlessNotExist(id, Base, context);
+    final.issuer = caller
+    final.currentOwner = caller
+    final.type = base.type as BaseType || BaseType.mixed
+    final.symbol = base.symbol.trim()
+    final.metadata = base.metadata
+
```
Comment (vikiival) [inline src/mappings/v2/base.ts:27]: missing impl of parts

File: src/mappings/v2/create.ts [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    }
+    
+    await context.store.save(final).then(() => {
+      success(OPERATION, final.id)
+    })
+  } catch (e) {
+    error(e, OPERATION, JSON.stringify(collection))
+  }
+}
+
```
[the remark this answers is not in the document]
Comment (vikiival) [reply src/mappings/v2/create.ts:63]: No-but feel free to do a PR 🤗

File: src/mappings/v2/equip.ts (hunk 10 above)
Comment (vikiival) [reply src/mappings/v2/equip.ts:27]: most probably not I'll plan to check critical issues in this comment but most probably merge as is

File: src/mappings/v2/equippable.ts (hunk 11 above)
Comment (vikiival) [reply src/mappings/v2/equippable.ts:29]: most probably not I'll plan to check critical issues in this comment but most probably merge as is

File: src/mappings/v2/index.ts [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+        break
+      case Interaction.LOCK:
+        await lockCollection(context)
+        break
+      case Interaction.DESTROY:
+      case Interaction.THEMEADD:  
+      case Interaction.EQUIPPABLE:
+      case Interaction.SETPROPERTY:
+      case Interaction.EQUIP:
+        pending(event, `::${base.blockNumber}::${base.value}`)
```
Comment (vikiival) [inline src/mappings/v2/index.ts:68]: all these are missing

File: src/mappings/v2/lock.ts [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+
+  try {
+    const { value, caller } = unwrap(context, getInteraction);
+    interaction = value
+    plsBe(withMeta, interaction)
+    const collection = await get<CollectionEntity>(context.store, CollectionEntity, interaction.id)
+    plsBe<CollectionEntity>(real, collection)
+    isOwnerOrElseError(collection, caller)
+    // TODO: implement lock
+    collection.max = Number(interaction.value)
```
Comment (vikiival) [inline src/mappings/v2/lock.ts:25]:
```ts
max = collection.supply 
```

File: src/mappings/v2/mint.ts [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+import { handleMetadata } from '../shared/metadata'
+import logger, { error, success } from '../utils/logger'
+import {
+  Action, Context, getNftId, Optional
+} from '../utils/types'
+import { getCreateToken } from './getters'
+
+const OPERATION = Action.MINT
+
+// TODO: MINT IS NOT CORRECTLY IMPLEMENTED
```
Comment (vikiival) [inline src/mappings/v2/mint.ts:22]: Oh so `recipient` could be either account or NFT, 
need to check this

Comment (vikiival) [reply src/mappings/v2/mint.ts:22]: const isRecipientNFT = !isDummyAddress(recipient)

File: src/mappings/v2/send.ts [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+    success(OPERATION, `${nft.id} to ${interaction.value}`)
+    await context.store.save(nft)
+    await createEvent(nft, Action.SEND, { blockNumber, caller, timestamp, version }, interaction.value || '', context.store, originalOwner)
+  } catch (e) {
+    error(e, OPERATION, JSON.stringify(interaction))
+  }
+}
+
+
+// current owner on nested should indicate the owner of the root NFT
```
Comment (vikiival) [inline src/mappings/v2/send.ts:59]:
```suggestion

```

File: src/mappings/v2/send.ts [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+    isMoreTransferable(nft, blockNumber)
+    const originalOwner = nft.currentOwner ?? undefined
+    const recipient = interaction.value || ''
+
+    const isRecipientNFT = !isDummyAddress(recipient)
+
+    nft.price = BigInt(0)
+    nft.updatedAt = timestamp
+
+
```
Comment (vikiival) [inline src/mappings/v2/send.ts:35]:
```suggestion
// current owner on nested should indicate the owner of the root NFT
```

File: src/mappings/v2/setproperty.ts (hunk 12 above)
Comment (vikiival) [reply src/mappings/v2/setproperty.ts:20]: 99.99% not :)
- need new property entity

File: src/model/generated/base.model.ts [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+    @Column_("text", {nullable: false})
+    currentOwner!: string
+
+    @Index_()
+    @ManyToOne_(() => MetadataEntity, {nullable: true})
+    meta!: MetadataEntity | undefined | null
+
+    @Column_("text", {nullable: true})
+    metadata!: string | undefined | null
+}
```
Comment (vikiival) [inline src/model/generated/base.model.ts:32]: missing parts

File: src/processor.ts [hunk 22]
```diff
@@ -5,8 +5,12 @@ import * as mappings from './mappings'
 
 const processor = new SubstrateProcessor(new Database())
 
+const STARTING_BLOCK = 10269144; // 8788586
+const ENDING_BLOCK = undefined; // 16261119;
```
Comment (vikiival) [inline src/processor.ts:9]:
```suggestion

const STARTING_BLOCK_V1 = 5756453;
const STARTING_BLOCK = 10269144; // 8788586
const ENDING_BLOCK = undefined; // 16261119;
```

File: src/server-extension/query/nft.ts [hunk 23]
```diff
@@ -37,3 +37,15 @@ WHERE
     AND e.timestamp >= NOW() - INTERVAL '${dateRange}'
 ORDER BY
     e.timestamp`
+
+
+export const rootOwnerQuery = `with recursive part as (
```
Comment (vikiival) [inline src/server-extension/query/nft.ts:42]: Luckily find @kremso's PDT presentations

(thanks for teaching that ^-^ )

**Review by vikiival — COMMENTED** [review]
That's a lot

**Comment by yangwao** [comment]
👀

File: src/mappings/shared/buy.ts (hunk 4 above)
Comment (vikiival) [reply src/mappings/shared/buy.ts]: a4e71ad877084ed7e491b6abc2f6a17ab92494b3

File: src/mappings/shared/emote.ts (hunk 6 above)
Comment (vikiival) [reply src/mappings/shared/emote.ts:16]: https://github.com/kodadot/rubick/issues/215

File: src/mappings/v2/create.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (vikiival) [reply src/mappings/v2/create.ts:63]: https://github.com/kodadot/rubick/issues/216

File: src/mappings/v2/equippable.ts (hunk 11 above)
Comment (vikiival) [reply src/mappings/v2/equippable.ts:29]: https://github.com/kodadot/rubick/issues/217

File: src/mappings/v2/lock.ts (hunk 17 above)
Comment (vikiival) [reply src/mappings/v2/lock.ts:25]: https://github.com/kodadot/rubick/issues/218

File: src/mappings/v2/mint.ts (hunk 18 above)
Comment (vikiival) [reply src/mappings/v2/mint.ts:22]: https://github.com/kodadot/rubick/issues/219

File: src/mappings/v2/create.ts (hunk 13 above)
Comment (vikiival) [reply src/mappings/v2/create.ts:16]: https://github.com/kodadot/rubick/issues/221

File: src/mappings/v2/base.ts (hunk 14 above)
Comment (vikiival) [reply src/mappings/v2/base.ts:27]: https://github.com/kodadot/rubick/issues/210

File: src/mappings/v2/equip.ts (hunk 10 above)
Comment (vikiival) [reply src/mappings/v2/equip.ts:27]: https://github.com/kodadot/rubick/issues/217

File: src/mappings/utils/entity.ts [hunk 24]
```diff
@@ -63,3 +67,16 @@ export function create<T extends EntityWithId>(
   Object.assign(entity, init)
   return entity
 }
+
+
+// Specific utils
```
[the remark this answers is not in the document]
Comment (vikiival) [reply src/mappings/utils/entity.ts]: It says that it was added because of RMRK2

File: src/mappings/utils/types.ts (hunk 8 above)
Comment (vikiival) [reply src/mappings/utils/types.ts:89]: https://github.com/kodadot/rubick/issues/222

File: src/mappings/shared/mint.ts (hunk 7 above)
Comment (vikiival) [reply src/mappings/shared/mint.ts:40]: fixed in ec543cceb3db6c711d87fa4dccb20cb450361fb2

**Comment by vikiival** [comment]
Other issues are tracked in:
- #188

**Pull request merged**